File: wearable/src/main/java/com/google/android/apps/muzei/datalayer/ActivateMuzeiIntentService.java
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.datalayer;

import android.app.IntentService;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.ResultReceiver;
import android.preference.PreferenceManager;
import android.service.notification.StatusBarNotification;
import android.support.annotation.RequiresApi;
import android.support.v4.app.NotificationCompat;
import android.support.v4.content.ContextCompat;
import android.support.wearable.activity.ConfirmationActivity;
import android.text.TextUtils;
import android.util.Log;
import android.widget.Toast;

import com.google.android.gms.tasks.Tasks;
import com.google.android.gms.wearable.CapabilityClient;
import com.google.android.gms.wearable.MessageClient;
import com.google.android.gms.wearable.Node;
import com.google.android.gms.wearable.Wearable;
import com.google.android.wearable.intent.RemoteIntent;
import com.google.android.wearable.playstore.PlayStoreAvailability;
import com.google.firebase.analytics.FirebaseAnalytics;

import net.nurik.roman.muzei.R;

import java.io.IOException;
import java.util.Set;
import java.util.TreeSet;
import java.util.concurrent.ExecutionException;

public class ActivateMuzeiIntentService extends IntentService {
    private static final String TAG = "ActivateMuzeiService";
    private static final String NOTIFICATION_CHANNEL = "activate_muzei";
    private static final int INSTALL_NOTIFICATION_ID = 3113;
    private static final int ACTIVATE_NOTIFICATION_ID = 3114;
    private static final String ACTIVATE_MUZEI_NOTIF_SHOWN_PREF_KEY = "ACTIVATE_MUZEI_NOTIF_SHOWN";
    private static final String ACTION_MARK_NOTIFICATION_READ =
            "com.google.android.apps.muzei.action.NOTIFICATION_DELETED";
    private static final String ACTION_REMOTE_INSTALL_MUZEI =
            "com.google.android.apps.muzei.action.REMOTE_INSTALL_MUZEI";

    public static void maybeShowActivateMuzeiNotification(Context context) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        if (preferences.getBoolean(ACTIVATE_MUZEI_NOTIF_SHOWN_PREF_KEY, false)) {
            return;
        }
        NotificationManager notificationManager = (NotificationManager)
                context.getSystemService(NOTIFICATION_SERVICE);
        StatusBarNotification[] notifications = notificationManager.getActiveNotifications();
        boolean hasInstallNotification = false;
        boolean hasActivateNotification = false;
        for (StatusBarNotification notification : notifications) {
            if (notification.getId() == INSTALL_NOTIFICATION_ID) {
                hasInstallNotification = true;
            } else if (notification.getId() == ACTIVATE_NOTIFICATION_ID) {
                hasActivateNotification = true;
            }
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            createNotificationChannel(context);
        }
        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, NOTIFICATION_CHANNEL);
        builder.setSmallIcon(R.drawable.ic_stat_muzei)
                .setColor(ContextCompat.getColor(context, R.color.notification))
                .setPriority(NotificationCompat.PRIORITY_MAX)
                .setDefaults(NotificationCompat.DEFAULT_VIBRATE)
                .setAutoCancel(true)
                .setContentTitle(context.getString(R.string.activate_title));
        Intent deleteIntent = new Intent(context, ActivateMuzeiIntentService.class);
        deleteIntent.setAction(ACTION_MARK_NOTIFICATION_READ);
        builder.setDeleteIntent(PendingIntent.getService(context, 0, deleteIntent, 0));
        // Check if the Muzei main app is installed
        CapabilityClient capabilityClient = Wearable.getCapabilityClient(context);
        Set<Node> nodes = new TreeSet<>();
        try {
            nodes = Tasks.await(capabilityClient.getCapability(
                    "activate_muzei", CapabilityClient.FILTER_ALL)).getNodes();
        } catch (ExecutionException|InterruptedException e) {
            Log.e(TAG, "Error getting all capability info", e);
        }
        if (nodes.isEmpty()) {
            if (hasInstallNotification) {
                // No need to repost the notification
                return;
            }
            // Send an install Muzei notification
            if (PlayStoreAvailability.getPlayStoreAvailabilityOnPhone(context)
                    != PlayStoreAvailability.PLAY_STORE_ON_PHONE_AVAILABLE) {
                builder.setContentText(context.getString(R.string.activate_no_play_store));
                FirebaseAnalytics.getInstance(context).logEvent("activate_notif_no_play_store", null);
            } else {
                builder.setContentText(context.getString(R.string.activate_install_muzei));
                Intent installMuzeiIntent = new Intent(context, ActivateMuzeiIntentService.class);
                installMuzeiIntent.setAction(ACTION_REMOTE_INSTALL_MUZEI);
                PendingIntent pendingIntent = PendingIntent.getService(context, 0, installMuzeiIntent, 0);
                builder.addAction(new NotificationCompat.Action.Builder(R.drawable.open_on_phone,
                        context.getString(R.string.activate_install_action), pendingIntent)
                        .extend(new NotificationCompat.Action.WearableExtender()
                                .setHintDisplayActionInline(true)
                                .setAvailableOffline(false))
                        .build());
                builder.extend(new NotificationCompat.WearableExtender()
                        .setContentAction(0));
                FirebaseAnalytics.getInstance(context).logEvent("activate_notif_play_store", null);
            }
            notificationManager.notify(INSTALL_NOTIFICATION_ID, builder.build());
            return;
        }
        // else, Muzei is installed on the phone/tablet, but not activated
        if (hasInstallNotification) {
            // Clear any install Muzei notification
            notificationManager.cancel(INSTALL_NOTIFICATION_ID);
        }
        if (hasActivateNotification) {
            // No need to repost the notification
            return;
        }
        String nodeName = nodes.iterator().next().getDisplayName();
        builder.setContentText(context.getString(R.string.activate_enable_muzei, nodeName));
        Intent launchMuzeiIntent = new Intent(context, ActivateMuzeiIntentService.class);
        PendingIntent pendingIntent = PendingIntent.getService(context, 0, launchMuzeiIntent, 0);
        builder.addAction(new NotificationCompat.Action.Builder(R.drawable.open_on_phone,
                context.getString(R.string.activate_action, nodeName), pendingIntent)
                .extend(new NotificationCompat.Action.WearableExtender()
                        .setHintDisplayActionInline(true)
                        .setAvailableOffline(false))
                .build());
        Bitmap background = null;
        try {
            background = BitmapFactory.decodeStream(context.getAssets().open("starrynight.jpg"));
        } catch (IOException e) {
            Log.e(TAG, "Error reading default background asset", e);
        }
        builder.extend(new NotificationCompat.WearableExtender()
                .setContentAction(0)
                .setBackground(background));
        FirebaseAnalytics.getInstance(context).logEvent("activate_notif_installed", null);
        notificationManager.notify(ACTIVATE_NOTIFICATION_ID, builder.build());
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private static void createNotificationChannel(Context context) {
        NotificationManager notificationManager = context.getSystemService(NotificationManager.class);
        NotificationChannel channel = new NotificationChannel(NOTIFICATION_CHANNEL,
                context.getString(R.string.activate_channel_name),
                NotificationManager.IMPORTANCE_HIGH);
        channel.enableVibration(true);
        notificationManager.createNotificationChannel(channel);
    }

    public static void clearNotifications(Context context) {
        NotificationManager notificationManager = (NotificationManager)
                context.getSystemService(NOTIFICATION_SERVICE);
        notificationManager.cancel(INSTALL_NOTIFICATION_ID);
        notificationManager.cancel(ACTIVATE_NOTIFICATION_ID);
    }

    public ActivateMuzeiIntentService() {
        super(TAG);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
        String action = intent.getAction();
        if (TextUtils.equals(action, ACTION_MARK_NOTIFICATION_READ)) {
            preferences.edit().putBoolean(ACTIVATE_MUZEI_NOTIF_SHOWN_PREF_KEY, true).apply();
            return;
        } else if (TextUtils.equals(action, ACTION_REMOTE_INSTALL_MUZEI)) {
            Intent remoteIntent =
                    new Intent(Intent.ACTION_VIEW)
                            .addCategory(Intent.CATEGORY_BROWSABLE)
                            .setData(Uri.parse("market://details?id=" + getPackageName()));
            RemoteIntent.startRemoteActivity(this, remoteIntent, new ResultReceiver(new Handler()) {
                @Override
                protected void onReceiveResult(int resultCode, Bundle resultData) {
                    if (resultCode == RemoteIntent.RESULT_OK) {
                        FirebaseAnalytics.getInstance(ActivateMuzeiIntentService.this)
                                .logEvent("activate_notif_install_sent", null);
                        preferences.edit().putBoolean(ACTIVATE_MUZEI_NOTIF_SHOWN_PREF_KEY,
                                true).apply();
                    } else {
                        Toast.makeText(ActivateMuzeiIntentService.this,
                                R.string.activate_install_failed, Toast.LENGTH_SHORT).show();
                    }
                }
            });
            return;
        }
        // else -> Open on Phone action
        CapabilityClient capabilityClient = Wearable.getCapabilityClient(this);
        Set<Node> nodes = new TreeSet<>();
        try {
            nodes = Tasks.await(capabilityClient.getCapability(
                    "activate_muzei", CapabilityClient.FILTER_REACHABLE)).getNodes();
        } catch (ExecutionException|InterruptedException e) {
            Log.e(TAG, "Error getting reachable capability info", e);
        }
        if (nodes.isEmpty()) {
            Toast.makeText(this, R.string.activate_failed, Toast.LENGTH_SHORT).show();
        } else {
            FirebaseAnalytics.getInstance(this).logEvent("activate_notif_message_sent", null);
            // Show the open on phone animation
            Intent openOnPhoneIntent = new Intent(this, ConfirmationActivity.class);
            openOnPhoneIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            openOnPhoneIntent.putExtra(ConfirmationActivity.EXTRA_ANIMATION_TYPE,
                    ConfirmationActivity.OPEN_ON_PHONE_ANIMATION);
            startActivity(openOnPhoneIntent);
            // Clear the notification
            NotificationManager notificationManager = (NotificationManager)
                    getSystemService(NOTIFICATION_SERVICE);
            notificationManager.cancel(INSTALL_NOTIFICATION_ID);
            preferences.edit().putBoolean(ACTIVATE_MUZEI_NOTIF_SHOWN_PREF_KEY, true).apply();
            // Send the message to the phone to open Muzei
            MessageClient messageClient = Wearable.getMessageClient(this);
            for (Node node : nodes) {
                try {
                    Tasks.await(messageClient.sendMessage(node.getId(),
                            "notification/open", null));
                } catch (ExecutionException|InterruptedException e) {
                    Log.w(TAG, "Unable to send Activate Muzei message to " + node.getId(), e);
                }
            }
        }
    }
}

File: wearable/src/main/java/com/google/android/apps/muzei/datalayer/ActivateMuzeiIntentService.kt
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.datalayer

import android.app.IntentService
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.ResultReceiver
import android.preference.PreferenceManager
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.app.TaskStackBuilder
import androidx.core.content.ContextCompat
import androidx.core.content.edit
import com.google.android.apps.muzei.ChooseProviderActivity
import com.google.android.apps.muzei.util.toast
import com.google.android.gms.wearable.CapabilityClient
import com.google.android.gms.wearable.Node
import com.google.android.gms.wearable.Wearable
import com.google.android.wearable.intent.RemoteIntent
import com.google.android.wearable.playstore.PlayStoreAvailability
import com.google.firebase.analytics.FirebaseAnalytics
import kotlinx.coroutines.tasks.await
import net.nurik.roman.muzei.R
import java.util.TreeSet
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeoutException

class ActivateMuzeiIntentService : IntentService(TAG) {

    companion object {
        private const val TAG = "ActivateMuzeiService"
        private const val NOTIFICATION_CHANNEL = "activate_muzei"
        private const val INSTALL_NOTIFICATION_ID = 3113
        private const val ACTIVATE_NOTIFICATION_ID = 3114
        private const val INSTALL_PENDING_PREF_KEY = "INSTALL_PENDING"
        private const val ACTION_MARK_NOTIFICATION_READ = "com.google.android.apps.muzei.action.NOTIFICATION_DELETED"
        private const val ACTION_REMOTE_INSTALL_MUZEI = "com.google.android.apps.muzei.action.REMOTE_INSTALL_MUZEI"

        fun hasPendingInstall(context: Context) =
                PreferenceManager.getDefaultSharedPreferences(context)
                        .getBoolean(INSTALL_PENDING_PREF_KEY, false)

        fun resetPendingInstall(context: Context) {
            PreferenceManager.getDefaultSharedPreferences(context).edit {
                putBoolean(INSTALL_PENDING_PREF_KEY, false)
            }
        }

        suspend fun checkForPhoneApp(context: Context) {
            val node: Node?
            try {
                node = getNode(context, CapabilityClient.FILTER_ALL)
            } catch (e: TimeoutException) {
                // Google Play services is out of date, nothing more we can do
                return
            }
            if (node != null) {
                // Muzei's phone app is installed, allow use of the DataLayerArtProvider
                context.packageManager.setComponentEnabledSetting(
                        ComponentName(context, DataLayerArtProvider::class.java),
                        PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
                        PackageManager.DONT_KILL_APP)

                // Ask if they want to use the DataLayerArtProvider
                sendEnableNotification(context, node)
            } else {
                // Muzei's phone app isn't installed, disallow use of the DataLayerArtProvider
                context.packageManager.setComponentEnabledSetting(
                        ComponentName(context, DataLayerArtProvider::class.java),
                        PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                        PackageManager.DONT_KILL_APP)

                // Check whether they can even install Muzei on their phone
                if (PlayStoreAvailability.getPlayStoreAvailabilityOnPhone(context) ==
                        PlayStoreAvailability.PLAY_STORE_ON_PHONE_AVAILABLE) {
                    // They can install Muzei. Mark that we're about to ask them
                    // since they might not click through the notification but
                    // instead go to their phone manually
                    PreferenceManager.getDefaultSharedPreferences(context).edit {
                        putBoolean(INSTALL_PENDING_PREF_KEY, true)
                    }

                    // Now ask them if they want to install Muzei on their phone
                    sendInstallNotification(context)
                }
            }
        }

        private fun sendEnableNotification(context: Context, node: Node) {
            val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
                    ?: return
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                createNotificationChannel(context)
            }
            val builder = NotificationCompat.Builder(context, NOTIFICATION_CHANNEL)
            builder.setSmallIcon(R.drawable.ic_stat_muzei)
                    .setColor(ContextCompat.getColor(context, R.color.notification))
                    .setAutoCancel(true)
                    .setContentTitle(context.getString(R.string.datalayer_enable_title))
                    .setContentText(context.getString(R.string.datalayer_enable_text,
                            node.displayName))
                    .setContentIntent(TaskStackBuilder.create(context)
                            .addNextIntentWithParentStack(Intent(context, ChooseProviderActivity::class.java))
                            .getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT))
            FirebaseAnalytics.getInstance(context).logEvent("activate_notif_installed", null)
            notificationManager.notify(ACTIVATE_NOTIFICATION_ID, builder.build())
        }

        private fun sendInstallNotification(context: Context) {
            val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
                    ?: return
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                createNotificationChannel(context)
            }
            val builder = NotificationCompat.Builder(context, NOTIFICATION_CHANNEL)
            builder.setSmallIcon(R.drawable.ic_stat_muzei)
                    .setColor(ContextCompat.getColor(context, R.color.notification))
                    .setAutoCancel(true)
                    .setContentTitle(context.getString(R.string.datalayer_install_title))
                    .setContentText(context.getString(R.string.datalayer_install_text))
            val deleteIntent = Intent(context, ActivateMuzeiIntentService::class.java).apply {
                action = ACTION_MARK_NOTIFICATION_READ
            }
            builder.setDeleteIntent(PendingIntent.getService(context, 0, deleteIntent, 0))

            val installMuzeiIntent = Intent(context, ActivateMuzeiIntentService::class.java).apply {
                action = ACTION_REMOTE_INSTALL_MUZEI
            }
            val pendingIntent = PendingIntent.getService(context, 0, installMuzeiIntent, 0)
            builder.addAction(NotificationCompat.Action.Builder(R.drawable.open_on_phone,
                    context.getString(R.string.datalayer_install_action), pendingIntent)
                    .extend(NotificationCompat.Action.WearableExtender()
                            .setHintDisplayActionInline(true)
                            .setAvailableOffline(false))
                    .build())
            FirebaseAnalytics.getInstance(context).logEvent("activate_notif_play_store", null)
            notificationManager.notify(INSTALL_NOTIFICATION_ID, builder.build())
        }

        @RequiresApi(Build.VERSION_CODES.O)
        private fun createNotificationChannel(context: Context) {
            val notificationManager = NotificationManagerCompat.from(context)
            val channel = NotificationChannel(NOTIFICATION_CHANNEL,
                    context.getString(R.string.datalayer_activate_channel_name),
                    NotificationManager.IMPORTANCE_DEFAULT)
            notificationManager.createNotificationChannel(channel)
        }

        @Throws(TimeoutException::class)
        private suspend fun getNode(context: Context, capability: Int): Node? {
            val capabilityClient = Wearable.getCapabilityClient(context)
            val nodes: Set<Node> = try {
                capabilityClient.getCapability("activate_muzei", capability).await().nodes
            } catch (e: ExecutionException) {
                Log.e(TAG, "Error getting all capability info", e)
                TreeSet()
            } catch (e: InterruptedException) {
                Log.e(TAG, "Error getting all capability info", e)
                TreeSet()
            }
            return nodes.firstOrNull()
        }
    }

    override fun onHandleIntent(intent: Intent?) {
        when (intent?.action) {
            ACTION_MARK_NOTIFICATION_READ -> {
                resetPendingInstall(this)
            }
            ACTION_REMOTE_INSTALL_MUZEI -> {
                val remoteIntent = Intent(Intent.ACTION_VIEW)
                        .addCategory(Intent.CATEGORY_BROWSABLE)
                        .setData(Uri.parse("market://details?id=$packageName"))
                RemoteIntent.startRemoteActivity(this, remoteIntent, object : ResultReceiver(Handler(Looper.getMainLooper())) {
                    override fun onReceiveResult(resultCode: Int, resultData: Bundle?) {
                        if (resultCode == RemoteIntent.RESULT_OK) {
                            FirebaseAnalytics.getInstance(this@ActivateMuzeiIntentService)
                                    .logEvent("activate_notif_install_sent", null)
                        } else {
                            toast(R.string.datalayer_install_failed)
                        }
                    }
                })
            }
        }
    }
}
